PHP 8.0 migration guide

cmb69 opened the pull request:
This is a follow-up to PR #164 which I've screwed up unrepairably.

nikic commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+   <listitem>
+    <para>
+     The <parameter>flag</parameter> parameter of <function>ob_implicit_flush</function> was changed
+     to accept bool values rather than int.
```
```suggestion
     to accept a bool rather than an int.
```
By the way, we have quite a few more int->bool changes, but have not kept track of them. We'll probably find out when updating signatures :)

cmb69 replied:
ACK :)

nikic commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+   </listitem>
+   <listitem>
+    <para>
+     The <varname>ZipArchive::lastId</varname> property to get index value of last added entry has been added.
```
```suggestion
     The <varname>ZipArchive::lastId</varname> property to get the index value of last added entry has been added.

```

nikic commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+   <listitem>
+    <para>
+     The <literal>'remove_path'</literal> option of <methodname>ZipArchive::addGlob</methodname> and
+     <methodname>ZipArchive::addPattern</methodname> is now treated as arbitrary string prefix (for
```
```suggestion
     <methodname>ZipArchive::addPattern</methodname> is now treated as an arbitrary string prefix (for
```

nikic commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+    <para>
+     The <literal>'remove_path'</literal> option of <methodname>ZipArchive::addGlob</methodname> and
+     <methodname>ZipArchive::addPattern</methodname> is now treated as arbitrary string prefix (for
+     consistency with the <literal>'add_path'</literal> option), whereas formerly it was treated as
```
```suggestion
     consistency with the <literal>'add_path'</literal> option), whereas formerly it was treated as a
```

cmb69 commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+   </listitem>
+   <listitem>
+    <para>
+     <literal>New</literal> and <literal>instanceof</literal> can now be used with arbitrary
```
No; that's just a sloppy conversion from [UPGRADING](https://github.com/php/php-src/blob/94938e4ee7b8a8b843b1deb56be1e60b6b74ce0b/UPGRADING#L719).

carusogabriel commented:
@sgolemon Wanna take a look as well?

carusogabriel commented on appendices/migration80/new-classes.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
```
New interfaces, like https://wiki.php.net/rfc/stringable, should we put it here, or in a new `new-interfaces.xml`?

cmb69 replied:
Stringable is currently in new-features. But see https://github.com/php/doc-en/pull/170#discussion_r516049289; not sure if we even should have a new-classes section.

cmb69 replied:
I have moved the new constants, functions and classes to the new features section, so this is resolved.

carusogabriel reviewed (commented):
Should we document that all the deprecated stuff from PHP 7.x were removed as well?

https://github.com/php/php-src/pull/3770

cmb69 commented:
> Should we document that all the deprecated stuff from PHP 7.x were removed as well?

At least most of that should already be there in https://github.com/php/doc-en/pull/170/files#diff-dc046224091f8e12cac61d0f1373afa026a8425c1dac1032440080786e77f047. Might be sensible to double-check.

cmb69 commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,401 @@
+   </listitem>
+   <listitem>
+    <para>
+     The <classname>WeakMap</classname> class has been added.
```
There is already the WeakMap class from PECL/weakref, and the class name above is linked to it. Guess we need a manual link instead. :(

cmb69 commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+    <para>
+     Private methods declared on a parent class no longer enforce any inheritance rules on the methods
+     of a child class. (with the exception of final private constructors)
+     <!-- RFC: https://wiki.php.net/rfc/inheritance_private_methods -->
```
Hmm, aren't the first three "Now allowed" comments the wrong way round? E.g. shouldn't "// Now allowed: Private method in parent is abstract." rather be "// Now allowed: Private method in parent is not abstract."

nikic replied:
Yeah, I can see how the formulation is confusing. Maybe it would be better to keep it more generic, like this?
```
class ChildClass extends ParentClass {
    // All of the following are now allowed, even though the modifiers aren't
    // the same as for the private methods in the parent class.
    public abstract function method1() {}
    public static function method2() {}
    public function method3() {}
    public function method4() {}
}
```

cmb69 commented on appendices/migration80/new-features.xml:
```diff
@@ -0,0 +1,402 @@
+    <para>
+     The <classname>ValueError</classname> class has been added.
+    </para>
+   </listitem>
```
Right! Not sure if we should have a separate "new classes" section.

nikic replied:
My vote goes to only having a single "new features" section. Or maybe separating it along "new language features" and "new library features" rather than specific symbol types.

cmb69 replied:
After further consideration, I agree.

cmb69 commented on appendices/migration80/incompatible.xml:
```diff
@@ -0,0 +1,1598 @@
+   <listitem>
+    <para>
+     Following the hash comment operator <literal>#</literal> immediately with an opening bracket is
+     not supported as a comment anymore since this syntax is now used for attributes.
```
Nice!

cmb69 commented on appendices/migration80/incompatible.xml:
```diff
@@ -0,0 +1,1598 @@
+    <para>
+     The GD extension now uses <classname>GdImage</classname> objects as the underlying data structure
+     for images, rather than &resource;s. These objects are completely opaque, i.e. they don't have
+     any methods.
```
Then I think it's better to drop is, because at least XMLWriter objects are not opaque.

cmb69 commented on appendices/migration80/incompatible.xml:
```diff
@@ -0,0 +1,1598 @@
+  <itemizedlist>
+   <listitem>
+    <para>
+     <function>odbc_connect</function> no longer reuses persistent connections.
```
If a persistent resource with the same parameters was already there, it would have been reused, so the description doesn't look wrong to me, but anyway, that "persistent" is superfluous.

cmb69 commented:
The union type documentation would be available with PR #174.

nikic reviewed (approved):
I think we should merge this and continue iterating in-tree.

Only thing I'd suggest is to maybe merge the Windows section into appropriate other places. E.g. the proc_open one into incompatible changes, and the other two into other changes).

cmb69 commented:
Thanks, @nikic! However, I have some concerns to merge this wrt. translations. Even if we add a note to the commit message that this is work in progress and should not be translated yet, translators may overlook or ignore that, and might have a hard time to follow up on changes. And I'm afraid there is the need for a lot of changes; for instance, I've just noticed that PhD apperently does not support linking to arbitrary elements, so it seems that we need `<sect3>` elements instead of the `<listitems>` (or patch PhD), and that would require restructuring (and maybe also some reordering). At least linking to the most prominent features/changes should be possible (e.g. to link from the release announcements).

cmb69 commented:
I have added subsections for the most relevant changes, integrated the Windows section into incompatible, added a few links to documentation available in the meantime, and applied as http://svn.php.net/viewvc?view=revision&revision=351500.
